lib/ansible/modules/cloud/vmware/vmware_host_facts.py

```diff
@@ -213,6 +234,9 @@ def get_system_facts(self):
 
 def main():
     argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        esxi_hostname=dict(type='str', required=False),
```

Comment:
This needs to be set to required. Otherwise, if you don't specify `esxi_hostname`, the first host will be used which doesn't make sense at all.

This should be:
```
argument_spec.update(
        cluster_name=dict(type='str', required=True),
    )
```

Reply:
Right, but you don't need to use it at all if you are connecting directly to an ESXi host so I don't think it should be required.  Is it required by any other modules?

The present behavior without this patch when connecting to a vCenter is to return the first host.  Perhaps the behavior should be to return all hosts if connecting to a vCenter without specifying `esxi_hostname`?  If I do this, I might as well implement `cluster_name` then too :-)

Reply:
I had a typo. It should be `esxi_hostname`. 

> Perhaps the behavior should be to return all hosts if connecting to a vCenter without specifying `esxi_hostname`? If I do this, I might as well implement `cluster_name` then too :-)

Like you already said. Looks like this doesn't work with ansible_facts.

Reply:
> Right, but you don't need to use it at all if you are connecting directly to an ESXi host so I don't think it should be required. Is it required by any other modules?

If you don't specify `esxi_hostname` and connect via vCenter, the first host is returned and the variables are messed up because for every host in the playbook run, the variables of the first host are used. Any conditional check doesn't work as expected anymore. This is especially dangerous if you execute the playbook without check mode.